File: day2/day2.java
import java.io.File;  
import java.io.FileNotFoundException;  
import java.util.Scanner;


public class day2 {
    

    public static void main(String[] args) {
      readFile();
    }

    static void readFile(){

        int counter1 = 0, counter2 = 0;
        try {
            File dict_file = new File("day2_passwords.txt");    	
            Scanner sc = new Scanner(dict_file);
            while (sc.hasNext()) {
              String password = sc.nextLine();

              if(validate(password))
                counter1++;
              if(validate(password,true))
                counter2++;
            }
            sc.close();
          } catch (FileNotFoundException e) {
            System.out.println("Invalid file");
            e.printStackTrace();
         }
    
         System.out.printf("Part 1: %d\nPart 2: %d", counter1, counter2);
    }

    static boolean validate(String s){
        return validate(s, false);
    }

    static boolean validate(String s, Boolean isPart2){
        int lowest = Integer.parseInt(s.substring(0,s.indexOf('-')));
        int highest = Integer.parseInt(s.substring(s.indexOf('-') + 1,s.indexOf(' ')));

       char c = s.charAt(s.indexOf(":") - 1);
       String testword = s.substring(s.indexOf(':') + 2);

       if(isPart2)
       return count2(testword, c, lowest, highest);
       else
       return count(testword, c, lowest, highest);

    }

    static boolean count(String s, char c, int low, int high){
        
        int counter = 0;
        
        for(char ch:s.toCharArray()){
            if(ch == c){
                counter++;
                if(counter > high)
                    return false;
            }
        }
        if(counter < low)
            return false;
        
        return true;
    }

    static boolean count2(String s, char c, int low, int high){
       return (s.charAt(low -1) == c && !(s.charAt(high -1 ) == c))
       || (!(s.charAt(low -1) == c) && (s.charAt(high - 1) == c));
        
    }
}

File: day2/day2.rb
arr = File.open("day2_passwords.txt", "r").map(&:chomp)
def check_if_valid(low, high, char, string)
    count = 0
    string.split('').each{|c| count += 1 if c == char}
    return true if count >= low.to_i && count <= high.to_i
    
end

def check_if_valid_2(low, high, char, string)
    return true if ((char == string[low.to_i] && char != string[high.to_i]) || 
                    (char != string[low.to_i] && char == string[high.to_i]))
end

counter_part_1, counter_part_2 = 0,0
arr.each do |x|  
    counter_part_1 += 1 if check_if_valid(x[0...x.index("-")], 
                    x[x.index("-")+1...x.index(" ")], 
                    x[x.index(':')-1], 
                    x[x.index(':')+1..])

    counter_part_2 += 1 if check_if_valid_2(x[0...x.index("-")], 
                    x[x.index("-")+1...x.index(" ")], 
                    x[x.index(':')-1], 
                    x[x.index(':')+1..])
end

puts "Part 1: #{counter_part_1} \nPart 2: #{counter_part_2}"
